### ISSUE TITLE
Support rendering of JSON arrays

### ISSUE THREAD

**Issue by nikolayhg** [angular-translate/angular-translate#769]
It would be great if one can render such arrays (key need for my application case):

```javascript
'content': [
    "Das ist ein langer Satz",
    "Das ist ein zweiter langer Satz",
],
'people' : [
  {
    'name': "John(DE)"
  },
  {
    'name': "Tim(DE)"
  }
]
```
with an expression like:

```html
<span translate="content[0]"></span>
<span translate="people[0].name"></span>
```

See (not working) example: http://plnkr.co/edit/u2Z3l2?p=preview

**Comment by adrianObel** [comment]
+1

**Comment by Swissbite** [comment]
It would also be great if I could get the array.

```javascript
'content': [
    "Das ist ein langer Satz",
    "Das ist ein zweiter langer Satz",
],
```
and in the controller:
```javascript
app.controller($translate, $scope) {
$translate('content').then(function(contentArray) {
  $scope.content = contentArray;
}))
}
```
and the corresponding view
```html
<ul>
<li ng-repeat='item in content'>item</li>
</ul>
```

**Comment by stryju** [comment]
@nikolayhg 

the keys are being "flattened", so just change your notation to "invalid" json, like so: http://plnkr.co/edit/K3pbBD7hJcjHJrjo1X4C?p=preview

---
tl;dr

use `array.index.property`, not `array[index]property`

---

i'd love, however, to get the full array, as in @Swissbite example - currently struggling with that...

**Comment by pfried** [comment]
+1 this would be awesome, we are having the use case of a list with a varying number of bullet points per language like

    [ "bullet point 1", "bullet point 2" ]

versus

    ["Listenpunkt 1"]

as they are translations (not per key of course) it would be great to have a syntax supporting `ng-repeat` and returning the translation  strings

**Comment by stryju** [comment]
as a workaround, you can:

####  1. use `translate-cloak` directive with trial and error... 
```html
<p ng-repeat="key in keys" translate="{{ ::key }}" translate-cloak="{{ ::key }}"></p>
```
now, just style the `.translate-cloak` class to hide the element
...not elegant, but "works"

#### 2. use `$translate` service and again - trial and error
```js
$translate( keys, function ( result ) {
  var translated = result.filter( function ( translation, index ) {
   return translation !== keys[ index ];
  });

  // use the translated in `for-each` or sth...
```
this one fails hard, if key SHOULD equal the translation...

**Comment by pfried** [comment]
@stryju  Could you create a plunker with a small demo? because i have problems understanding the syntax you used. I would like to try the first solution

**Comment by knalli** [comment]
Adding an example f0a6ca3 showing that arrays are supported and can be used already.

Closing this in forward.

**Comment by Swissbite** [comment]
@knalli I disagree. This may be a workaround. But it's not a full support of an array.
My idea was to support arrays without knowledge of its length. 
Currently there is no benefit of arrays in translation, it is handled as a normal Jason object and the developer has to know the length of the array.

Arrays would be perfect for lists if they are easy editable without the necessary to either edit the template or to handle the logic in the controller.

**Comment by knalli** [comment]
What is your proposed style?

**Comment by pfried** [comment]
I guess it would be the best if we could get to the array itself somehow.

Get the translation array:

    $scope.translations = $translate('keyOfTheArray');

Put it in the view

    <div ng-repeat="translation in translations">
      <p>{{ translation }}</p>
    </div>

The translate filter is used inside of an element. As one usually needs to repeat HTML Elements i do not think there is a use case other than accessing them by index. Maybe something like

    <p>{{ 'keyOfTheArray' | translate::1}}</p>

The translate directive could be extended using another attribute. But it will "break" the directive anyway as it is not intended for repeated elements:

    <ul translate="keyOfTheArray" translate-repeat="true">
      <li>{{ translation }}</li>
    </ul>

I think it would be the best if we could get the array in the controller and just use an `ng-repeat` for those non associative arrays.

**Comment by knalli** [comment]
Okay, I had not get this right. Currently everything is based on strings, but not sth else (like an array).

That feature is currently not to be planned; a contribution should cover

* service calls (both $translate() and $translate.instant()), incl. interpolation params
* a directive extension (i.e. translate-repeat) incl. interpolation params, values, defaults

Also there should be strategy regarding 

* what happens when a non array was requested with -repeat way?
* what happens when a array was requested with standard way?
* what happens with non existing value requested with -repeat? what happens with default?

**Comment by stryju** [comment]
maybe keeping the current logic + extending it with aliases would be an option?

---

in the "ideal world", this would be possible:

```js
{
  FOO: {
    BAR: {
      BAZ: [
        "xxx"
        "yyy",
        "zzz"
      ],
      QUUX: "abc",
      WUT: function( $scope, SomeService ) {
        return SomeService.serviceMethod( $scope );
      }
    }
  }
}
```

```html
<h1 translate="FOO.BAR.QUUX"></h1>
<ul ng-init="alias = ('FOO.BAR.BAZ'|translate)">
  <li ng-repeat="item in alias">{{ item }}
</ul>
<p translate="FOO.BAR.WUT"></p>
```

**Comment by knalli** [comment]
In the "ideal world" that isn't JSON anymore.

Perhaps we can introduce in the next major release some execution points for these extensions.

**Comment by stryju** [comment]
the `WUT` part isn't - true - but except for that, it is ;-)

i don't know how big of a performance hit would it be to have a key accessor be a structure crawler, like:

```js
function accessor( obj, key ) {
  var chain = key.replace( /\[\s*([\w_]+)\s*\]/g, '.$1' ).split( '.' );
  var chainLength = chain.length;
  var chainStep = 0;
  var alias = obj;

  console.log( chain );

  try {
    while ( alias !== void 0 && chainStep < chainLength && typeof alias !== 'string' ) {
      alias = alias.hasOwnProperty( chain[ chainStep ] ) ?
        alias[ chain[ chainStep++ ] ] :
        void 0;
    }
  } catch( e ) {
    console.warn( e );
    alias = void 0;
  }

  if ( typeof alias === 'string' && chainStep < chainLength ) {
    return void 0;
  }

  return alias;
}
```

then

```js

accessor( 'FOO' ) // { BAR: { BAZ : [ 'xxx', 'yyy', 'zzz', ], QUUX: 'abc' } }
accessor( 'FOO.BAR' ) // { BAZ : [ 'xxx', 'yyy', 'zzz', ], QUUX: 'abc' }
accessor( 'FOO[BAR]' ) // { BAZ : [ 'xxx', 'yyy', 'zzz', ], QUUX: 'abc' }
accessor( 'FOO.QUUX' ) // 'abc'
accessor( 'FOO.BAR.BAZ' ) // [ 'xxx', 'yyy', 'zzz', ]
accessor( 'FOO.BAR.BAZ.0' ) // 'xxx'
accessor( 'FOO.BAR.BAZ[0]' ) // 'xxx'
accessor( 'FOO.BAR.BAZ.0.1' ) // undefined
```

**Comment by stryju** [comment]
and to answer
>  what happens when a non array was requested with -repeat way?

```html
    <p>Hello {{name}}!</p>
    <ul>
      <li ng-repeat="x in name">{{ x }}</li>
    </ul>
```

output:

> Hello World!
> - W
> - o
> - r
> - l
> - d


http://plnkr.co/edit/yziOimu2KHUk9pB3JLm9?p=preview

**Comment by cabloo** [comment]
My solution to this:

JSON:

```json
{
    "ACTIONS": [
          "Test Action 1",
          "Test Action 2"
    ],
    "ACTION_COUNT": "2",
}
```

Template:

```jade
ul(ng-init="items = _.range(0, 'ACTION_COUNT' | translate)")
  li(
    ng-repeat="item in items track by $index"
    translate="ACTIONS.{{$index}}"
  )
```

Requires lodash/underscore available in `$scope`, e.g.

```
angular
    .module('app.core')
    .run(appRun);

  /**
   * @ngInject
   */
  function appRun($rootScope, _) {
    $rootScope._ = _;
  }
```

A directive that uses the appended `_COUNT` could make this whole thing transparent to the developer.

**Comment by dackmin** [comment]
+1

**Comment by dangjian** [comment]
+1

**Comment by kensplanet** [comment]
+1

**Comment by VAMSHI18** [comment]
Is this being supported now? (i.e, $translate loading array from resource JSON file)?

**Comment by knalli** [comment]
@VAMSHI18 Please scroll up starting with https://github.com/angular-translate/angular-translate/issues/769#issuecomment-92227400
